### DISCUSSION TITLE
How to expand SUP SubstanceGroups?

### DISCUSSION THREAD

**Discussion by someonetookmyname** [rdkit/rdkit#6439] [Q&A] [answered]
Hi,

I'm writing to inquire about how to expand SUP SubstanceGroups in a molecule?

For the V3000 mol spec, 'ESTATE=E' seems to indicate an expanded superatom (SubstanceGroup), and manually editing (outside of rdkit) V3000 mol data for a contracted superatom to have 'ESTATE=E' expands the superatom as expected.

V2000 mol data have a slightly different spec, but SubstanceGroup handling in rdkit seems to follow the V3000 spec.  That is, if I load a V3000 molecule into rdkit with an already-expanded superatom, GetProps('ESTATE') for the SubstanceGroup returns 'E', as expected.  For V2000 data, GetProps('ESTATE') for the SubstanceGroup also returns 'E' (ie. doesn't seem to use the 'SDS EXP' spec of V2000).

However, if I try to expand a contracted SubstanceGroup by SetProps('ESTATE', 'E'), the superatoms in an output mol block are not expanded.

That is, I have unsuccessfully tried:
```
     for sgs in Chem.GetMolSubstanceGroups(mol):
          if sgs.GetProp('TYPE')=='SUP':
               sgs.SetProp('ESTATE','E')
     print(Chem.MolToV3KMolBlock(mol))
```

Is it possible to expand SUP SubstanceGroups?

Thank you for any guidance.

Best,

Andrew

**Comment by greglandrum** [comment, marked as answer]
Hi @someonetookmyname. The RDKit has code to collapse abbreviations, but there's not currently anything there to expand them.
Here's an example which uses the built-in set of abbreviations, but the code should also work with SUP groups.
https://www.rdkit.org/docs/Cookbook.html#with-abbreviations

**Reply by someonetookmyname** [reply]
Thanks for clarifying, Greg.  I think an expand option could be useful, if possible.

In the meantime, in case anyone lands here, here is some code (granted, a bit tedious) that will do expansion by generating the V3000 mol block for a particular structure, doing low-level parsing and modification of the mol block to set the expanded state of abbreviations, and then reimporting the block.

```
from rdkit import Chem
import re
import sys
import copy

mols = Chem.SDMolSupplier(sys.argv[1], sanitize=False, removeHs=False)
for mol in mols:
    any_sup = False
    for sgs in Chem.GetMolSubstanceGroups(mol):
        if sgs.GetProp('TYPE')=='SUP':
            any_sup = True
            break
    if any_sup:
        mb = Chem.MolToMolBlock(mol, includeStereo=True, kekulize=True, forceV3000=True)
        mb_lines = mb.splitlines(True)
        index = 0
        in_sgroups = False
        while index < len(mb_lines):
            if re.search('^M  V30 BEGIN SGROUP', mb_lines[index]):
                in_sgroups = True
            elif re.search('^M  V30 END SGROUP', mb_lines[index]):
                in_sgroups = False
            elif in_sgroups==True and re.search('^M  V30 \d+ SUP', mb_lines[index]):
                found_estate = False
                found_end = False
                while found_end == False:
                    if re.search('ESTATE', mb_lines[index]):
                        found_estate = True
                    if re.search(' -$', mb_lines[index].rstrip()) is None:
                        found_end = True
                    else:
                        index+=1
                if found_estate == False:
                    mb_lines[index] = mb_lines[index].rstrip()+" -\nM  V30 ESTATE=E\n"
            index+=1
        expanded_mb = ''.join(mb_lines)
        xmol = Chem.MolFromMolBlock(expanded_mb, sanitize=False, removeHs=False)
        mol = copy.deepcopy(xmol)
```
